### PR TITLE
feat(APIv2): RHINENG-11773 filter failed rules by severity

### PR DESCRIPTION
**Pull request by romanblanco**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: spec/controllers/v2/test_results_controller_spec.rb [hunk 1]
```diff
@@ -72,6 +73,14 @@
         end.sort_by(&:id)
       end
 
+      let(:rule_result_data) do
+        %w(low medium high).each_with_object({}) do |level, obj|
+          obj["#{level}_severity_failing_rule".to_sym] = pw({
+            Faker::Internet.uuid => [level, 'fail']
```
Comment (skateman) [inline spec/controllers/v2/test_results_controller_spec.rb]: AFAIK you don't need the left side, so you can get rid of these objects completely. SImply define everything in the YAML.

**Review by skateman — CHANGES_REQUESTED** [review]
We can get to something like this:
```suggestion
        :additional_rule_results:
          - severity: 'high'
             result: 'pass'
          - severity: 'high'
             result: 'fail'
          - severity: 'low'
             result: 'fail'
```

File: app/models/v2/test_result.rb [hunk 2]
```diff
@@ -70,6 +70,15 @@ class TestResult < ApplicationRecord
       { conditions: systems.arel.where_sql.gsub(/^where /i, '') }
     end
 
+    searchable_by :failed_rule_severity, %i[in notin] do |_key, op, val|
+      values = val.split(',').map(&:strip)
```
[the remark this answers is not in the document]
Comment (romanblanco) [reply app/models/v2/test_result.rb:74]: We do need to separate the values here.

[the remark this answers is not in the document]
Comment (skateman) [reply app/models/v2/test_result.rb:74]: Yes, but do we need a `,` or isn't space enough? The rule of thumb is if there's space in the strings that you're searching for (see other examples).

[the remark this answers is not in the document]
Comment (romanblanco) [reply app/models/v2/test_result.rb:74]: There is no space in the string to split by.

File: app/models/v2/test_result.rb [hunk 3]
```diff
@@ -70,6 +70,15 @@ class TestResult < ApplicationRecord
       { conditions: systems.arel.where_sql.gsub(/^where /i, '') }
     end
 
+    searchable_by :failed_rule_severity, %i[in] do |_key, _op, val|
+      values = val.split(',').map(&:strip)
+      ids = ::V2::TestResult.joins(rule_results: :rule)
+                            .where(v2_rules: { severity: values }, rule_results: { result: 'fail' })
+                            .reselect(:id)
+
+      { conditions: "v2_test_results.id IN (#{ids.to_sql})" }
```
Comment (skateman) [inline app/models/v2/test_result.rb:79]: I know why you're bumping into the select cache: you don't need to select from the test results table at all. You can directly go to rule_results and join it with rules + select by `test_result_id`. That way you're joining one less table and you won't need the `reselect`

File: spec/fixtures/files/searchable/test_results_controller.yaml [hunk 4]
```diff
@@ -294,3 +294,33 @@
...
+
+- :name: "in search by failed_rule_severity"
+  :entities:
+    :found:
+      - :factory: :v2_test_result
+        :account: ${account}
+        :policy_id: ${policy.id}
+        :additional_rule_results:
+          - severity: 'high'
+            result: 'fail'
```
Comment (skateman) [inline spec/fixtures/files/searchable/test_results_controller.yaml:306]: Please add more than just a single rule result here + combine them with passing tests with higher severity to properly cover the DB lookup.

File: app/models/v2/test_result.rb [hunk 5]
```diff
@@ -70,6 +70,15 @@ class TestResult < ApplicationRecord
       { conditions: systems.arel.where_sql.gsub(/^where /i, '') }
     end
 
+    searchable_by :failed_rule_severity, %i[in] do |_key, _op, val|
```
Comment (skateman) [inline app/models/v2/test_result.rb:73]: If you do `in` then you'd also have to do `eq` for semantic compatibility.

File: app/models/v2/test_result.rb [hunk 6]
```diff
@@ -70,6 +70,15 @@ class TestResult < ApplicationRecord
       { conditions: systems.arel.where_sql.gsub(/^where /i, '') }
     end
 
+    searchable_by :failed_rule_severity, %i[eq in] do |_key, op, val|
+      values = ['IN'].include?(op) ? val.split : val
```
Comment (skateman) [inline app/models/v2/test_result.rb:74]: You don't need to deal with the operator at all, the `split` returns a single element array when no separator is detected. So you can always call `split`, see other existing examples.

**Review by skateman — CHANGES_REQUESTED** [review]
One minor issue.

File: spec/fixtures/files/searchable/test_results_controller.yaml [hunk 7]
```diff
@@ -294,3 +294,82 @@
...
+        :policy_id: ${policy.id}
+        :additional_rule_results:
+          - severity: 'low'
+            result: 'fail'
+      - :factory: :v2_test_result
+        :account: ${account}
+        :policy_id: ${policy.id}
+        :additional_rule_results:
+          - severity: 'high'
+            result: 'pass'
```
Comment (skateman) [inline spec/fixtures/files/searchable/test_results_controller.yaml:333]: Nit: use symbols

File: spec/factories/test_result.rb [hunk 8]
```diff
@@ -34,7 +34,23 @@
...
+      ev.additional_rule_results.each do |rr|
+        FactoryBot.create(
+          :v2_rule_result,
+          rule: FactoryBot.create(
+            :v2_rule,
+            security_guide: tr.tailoring.security_guide
+          ),
+          test_result_id: tr.id,
+          severity: rr['severity'],
+          result: rr['result']
```
Comment (skateman) [inline spec/factories/test_result.rb:51]: Nit: use symbols

**Review by skateman — COMMENTED** [review]
One last thing as you need to rebase anyway

**Review by skateman — APPROVED** [review]
The Seal of Approval

**Pull request merged**